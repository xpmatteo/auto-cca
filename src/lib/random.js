
export function randomShuffleArray(array) {
    for (let i = 0; i < array.length; i++) {
        const j = Math.floor(Math.random() * (array.length - i)) + i;
        [array[i], array[j]] = [array[j], array[i]];
    }
}

/**
 * @template T
 * @param {T[]} array
 * @returns {T}
 */
export function randomElement(array) {
    return array[Math.floor(Math.random() * array.length)];
}

const RANDOM_VALUES = [0.1, 0.3, 0.2, 0.7, 0.9, 0.6, 0.4, 0.8];
let nextRandom = 0;
export function fixedRandom() {
    return RANDOM_VALUES[nextRandom++ % 8];
}
