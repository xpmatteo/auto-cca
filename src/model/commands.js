import { DamageEvent, BattleBackEvent } from "./events.js";

export class MoveCommand {
    constructor(to, from) {
        this.to = to;
        this.from = from;
    }

    toString() {
        return `Move ${this.from} to ${this.to}`;
    }

    play(game) {
        game.moveUnit(this.to, this.from);
        game.markUnitSpent(game.unitAt(this.to));
        game.addMovementTrail(this.to, this.from);
    }
}

export class EndPhaseCommand {
    toString() {
        return `End phase`;
    }

    play(game) {
        game.endPhase();
    }
}

export class CloseCombatCommand {
    constructor(toHex, fromHex) {
        this.toHex = toHex;
        this.fromHex = fromHex;
    }

    toString() {
        return `Close Combat from ${this.fromHex} to ${this.toHex}`;
    }

    play(game) {
        const attackingUnit = game.unitAt(this.fromHex);
        if (!attackingUnit) {
            throw new Error(`No unit at ${this.fromHex}`);
        }
        const defendingUnit = game.unitAt(this.toHex);
        if (!defendingUnit) {
            throw new Error(`No unit at ${this.toHex}`);
        }
        if (attackingUnit.side === defendingUnit.side) {
            throw new Error(`Cannot attack own unit at ${this.toHex}`);
        }
        if (this.toHex === this.fromHex) {
            throw new Error(`Cannot attack self at ${this.toHex}`);
        }
        if (this.toHex.distance(this.fromHex) > 1) {
            throw new Error(`Cannot Close Combat with unit at ${this.toHex} from ${this.fromHex} (too far)`);
        }
        const diceResults = game.roll(attackingUnit.diceCount());
        const damage = defendingUnit.takeDamage(diceResults);
        return [
            new DamageEvent(this.toHex, damage, diceResults),
            new BattleBackEvent(this.fromHex, this.toHex, defendingUnit.diceCount())
        ]
    }
}

