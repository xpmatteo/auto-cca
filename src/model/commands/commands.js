const DISTANCE_VALUE_BACKOFF = 0.2;

// Optimization: precompute the values for the backoff function
function backoffValues(number) {
    let values = [];
    for (let i = 0; i < number; i++) {
        values.push(Math.pow(DISTANCE_VALUE_BACKOFF, i));
    }
    return values;
}
const BACKOFF_VALUES = backoffValues(17);

/*
    Commands are immutable objects representing the actions that can be taken by the player. 
    They are used by both the human player and the AI to execute a move.
*/

export function hexScore(enemyUnitStrength) {
    return 1000 / enemyUnitStrength;
}

function valueOfHex(hexToBeEvaluated, enemyUnitHex, enemyUnitStrength) {
    const distance = enemyUnitHex.distance(hexToBeEvaluated);
    return hexScore(enemyUnitStrength) * BACKOFF_VALUES[distance];
}

export function valueOfHexOverAllEnemyUnits(game, hexToBeEvaluated, friendlySide) {
    let bestValue = -Infinity;
    game.foreachUnit((otherUnit, otherUnitHex) => {
        if (otherUnit.side === friendlySide) {
            return;
        }
        const candidateBest = valueOfHex(otherUnitHex, hexToBeEvaluated, game.unitStrength(otherUnit));
        if (candidateBest > bestValue) {
            bestValue = candidateBest;
        }
    });
    return bestValue;
}


export class FlagResult {
    constructor(damage, retreatHexes) {
        this.damage = damage;
        this.retreats = retreatHexes;
    }

    static NO_EFFECT = new FlagResult(0, []);

    toString() {
        return `FlagResult(damage: ${this.damage}, retreatHexes: ${this.retreats})`;
    }

    static retreat(hexes) {
        return new FlagResult(0, hexes);
    }

    static damage(number) {
        return new FlagResult(number, []);
    }
}

export function handleFlags(flags, retreatHexesPerFlag, ignorableFlags, retreatPaths) {
    if (flags === 0) {
        return FlagResult.NO_EFFECT;
    }
    let retreatMin = (flags - ignorableFlags) * retreatHexesPerFlag;
    let retreatMax = flags * retreatHexesPerFlag;
    console.log(`handleFlags: retreatMin: ${retreatMin}, retreatMax: ${retreatMax}, retreatAvailable: ${(retreatPaths.maxDistance)}`);

    if (retreatMin >= retreatPaths.maxDistance) {
        // damage and possibly retreat; no choice except retreat as much as possible
        let damage = retreatMax - retreatPaths.maxDistance;
        let retreatHexes = retreatPaths.maxDistance === 0 ? [] : retreatPaths[retreatPaths.maxDistance];
        console.log("handleFlags: case (1)");
        return new FlagResult(damage, retreatHexes);
    }
    if (retreatMax > retreatPaths.maxDistance) {
        console.log("handleFlags: case (2)");
        return new FlagResult(0, []);
    }

    // no damage -- just retreat
    if (ignorableFlags === 0) {
        let retreatHexes = retreatPaths.maxDistance === 0 ? [] : retreatPaths[retreatPaths.maxDistance];
        console.log("handleFlags: case (3)");
        return new FlagResult(0, retreatHexes);
    }
    let retreatHexes;
    if (retreatPaths.maxDistance === 0) {
        retreatHexes = [];
    } else {
        retreatHexes = retreatPaths[retreatMin].concat(retreatPaths[retreatPaths.maxDistance]);
    }
    console.log("handleFlags: case (4)");
    return new FlagResult(0, retreatHexes);
}
