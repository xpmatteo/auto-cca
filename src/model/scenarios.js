import { hexOf } from "../lib/hexlib.js";
import GameStatus from "./game_status.js";
import { Side } from "./side.js";
import * as units from "./units.js";

export class Scenario {
    /**
     * @param side
     * @returns {Side}
     */
    opposingSide(side) {
        return side === this.sideNorth ? this.sideSouth : this.sideNorth;
    }

    /**
     * @param {Game} game
     * @returns {GameStatus}
     */
    gameStatus(game) {
        if (game.killedUnitsOfSide(Side.CARTHAGINIAN).length === this.pointsToWin) {
            return GameStatus.ROMAN_WIN;
        }
        if (game.killedUnitsOfSide(Side.ROMAN).length === this.pointsToWin) {
            return GameStatus.CARTHAGINIAN_WIN;
        }
        return GameStatus.ONGOING;
    }

    /**
     * @param {Board|Game} board
     */
    placeUnitsOn(board) {
        throw new Error("Subclass must implement");
    }
}

export class NullScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 1;
    commandSouth = 1;

    placeUnitsOn(board) {
    }

    gameStatus(board) {
        return GameStatus.ONGOING;
    }
}

export class MeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 3;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        // 1522
        board.placeUnit(hexOf(0, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(1, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(2, 5), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(3, 5), new units.RomanHeavyInfantry());

        // 6422
        // board.placeUnit(hexOf(1, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(2, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(3, 4), new units.RomanHeavyInfantry());
        // board.placeUnit(hexOf(4, 4), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(2, 3), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 3), new units.CarthaginianHeavyInfantry());
    }
}

export class OneToOneMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 1;
    commandNorth = 3;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(2, 2), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.RomanLightBowsInfantry());
    }
}

export class TwoOnTwoMeleeScenario extends Scenario {
    firstSide = Side.ROMAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.ROMAN;
    pointsToWin = 2;
    commandNorth = 4;
    commandSouth = 3;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(-1, 6), new units.RomanHeavyInfantry());
        board.placeUnit(hexOf(0, 6), new units.RomanHeavyInfantry());

        board.placeUnit(hexOf(0, 4), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(1, 4), new units.CarthaginianHeavyInfantry());
    }
}


export class AkragasScenario extends Scenario {
    firstSide = Side.SYRACUSAN;
    sideNorth = Side.CARTHAGINIAN;
    sideSouth = Side.SYRACUSAN;
    pointsToWin = 5;
    commandNorth = 5;
    commandSouth = 6;

    placeUnitsOn(board) {
        board.placeUnit(hexOf(1, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(2, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(3, 1), new units.CarthaginianHeavyInfantry());
        board.placeUnit(hexOf(4, 1), new units.CarthaginianMediumInfantry());
        board.placeUnit(hexOf(6, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(7, 1), new units.CarthaginianAuxiliaInfantry());
        board.placeUnit(hexOf(8, 1), new units.CarthaginianAuxiliaInfantry());

        board.placeUnit(hexOf(1, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(3, 2), new units.CarthaginianLightBowsInfantry());
        board.placeUnit(hexOf(7, 2), new units.CarthaginianLightInfantry());
        board.placeUnit(hexOf(9, 2), new units.CarthaginianHeavyChariot());
        board.placeUnit(hexOf(10, 1), new units.CarthaginianLightCavalry());

        function placeSyracusanUnit(hex, unit) {
            board.placeUnit(hex, unit);
            unit.side = Side.SYRACUSAN;
        }
        placeSyracusanUnit(hexOf(-2, 7), new units.RomanLightCavalry());
        placeSyracusanUnit(hexOf(-1, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(0, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(2, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(3, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(4, 7), new units.RomanHeavyInfantry());
        placeSyracusanUnit(hexOf(5, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(6, 7), new units.RomanAuxiliaInfantry());
        placeSyracusanUnit(hexOf(7, 7), new units.RomanMediumCavalry());

        placeSyracusanUnit(hexOf(0, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(3, 6), new units.RomanLightBowsInfantry());
        placeSyracusanUnit(hexOf(7, 6), new units.RomanLightInfantry());
    }
}

const SCENARIOS = {
    akragas: new AkragasScenario(),
    melee: new MeleeScenario(),
    oneToOneMelee: new OneToOneMeleeScenario(),
    twoOnTwoMelee: new TwoOnTwoMeleeScenario(),
}

export function makeScenario(name) {
    return SCENARIOS[name] || new AkragasScenario();
}
