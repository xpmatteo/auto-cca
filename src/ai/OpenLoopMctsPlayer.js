import { InteractiveGame } from "../interactive_game.js";
import { randomElement } from "../lib/random.js";
import { Command } from "../model/commands/commands.js";
import { Game } from "../model/game.js";
import OpenLoopNode from "./OpenLoopNode.js";
import { scoreMcts } from "./score.js";


const DEFAULT_EXPANSION_FACTOR = 1.4142135623730951;

export class OpenLoopMctsPlayer {
    static isSearching = false;

    constructor(args = {}) {
        this.args = args;
        this.args.expansionFactor = this.args.expansionFactor || DEFAULT_EXPANSION_FACTOR;
        this.args.iterations = this.args.iterations || 1000;
        this.args.playoutIterations = this.args.playoutIterations || 20;
        this.args.logfunction = this.args.logfunction || console.log;
        this.args.samplingExplorationChance = this.args.samplingExplorationChance || 0.01;
        this.args.note = this.args.note || "";
    }

    /**
     * @param {Game} game
     * @returns {Command[]}
     */
    decideMove(game) {
        const startTime = Date.now();
        this.args.logfunction(" ----- AI IS THINKING -----")
        if (game.validCommands().length === 1) {
            return [game.validCommands()[0]];
        }
        const iterations = (game.currentPhase.requiresDeepThought()) ?
            this.args.iterations : 1000;
        const rootNode = this.search(game.toGame(), iterations);
        const bestCommands = rootNode.bestCommands(game.currentSide);
        this.args.logfunction(showBestCommands(rootNode.bestCommands(game.currentSide)));
        this.args.logfunction(rootNode.shape());
        this.args.logfunction(rootNode.toString(2));
        //this.args.logfunction(rootNode.redundancy());
        // this.args.logfunction(rootNode.toString(7, 10));
        this.args.logfunction("Time taken: " + (Date.now() - startTime)/1000 + "s");
        return bestCommands;
    }

    /**
     * @param {Game} game
     * @param {number} iterations
     * @returns {OpenLoopNode}
     */
    search(game, iterations = this.args.iterations) {
        OpenLoopMctsPlayer.isSearching = true;
        const rootNode = new OpenLoopNode(game.currentSide);
        for (let i = 0; i < iterations; i++) {
            if (i % 10000 === 0) this.args.logfunction("Iteration " + i);
            this.iterate(game, rootNode);
        }
        OpenLoopMctsPlayer.isSearching = false;
        return rootNode;
    }

    /**
     * @param {Game} game
     * @param {OpenLoopNode} rootNode
     */
    iterate(game, rootNode) {
        const [newGame, leafNode] = this._select(game, rootNode);
        const score = this._simulate(newGame);
        leafNode.backPropagate(score, newGame.currentSide);
    }

    /**
     * @param {Game} game
     * @param {OpenLoopNode} node
     * @returns {[Game,OpenLoopNode]}
     * @private
     */
    _select(game, node) {
        while (!game.isTerminal()) {
            if (node.childrenSize === 0) {
                return node.bestUctChild(game, this.args.expansionFactor);
            } else {
                [game, node] = node.bestUctChild(game, this.args.expansionFactor);
            }
        }
        // game is terminal
        return [game, node];
    }

    /**
     * @param {Game} game
     * @returns {number}
     * @private
     */
    _simulate(game) {
        const clone = game.clone();
        for (let i = 0; i < this.args.playoutIterations && !clone.isTerminal(); i++) {
            const commands = clone.validCommands();
            const command = randomElement(commands);
            if (command)
                clone.executeCommand(command);
        }
        const number = scoreMcts(clone, game.currentSide);
        // return number;
        if (number === 0) {
            return 0;
        }
        if (number < 0) {
            return -1;
        }
        if (number > 0) {
            return 1;
        }
    }

    toString() {
        return `MctsPlayer(${this.args.iterations}, ${this.args.playoutIterations}, ${this.args.samplingExplorationChance}, ${this.args.note}))`;
    }
}

function showBestCommands(comands) {
    return "Best commands: \n - " + comands.join("\n - ") + "\nEnd best commands";
}
