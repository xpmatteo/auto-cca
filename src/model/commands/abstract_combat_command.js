import { AdvanceAfterCombatPhase } from "../phases/advance_after_combat_phase.js";
import { DamageEvent, UnitKilledEvent } from "../events.js";
import * as dice from "../dice.js";
import { RetreatPhase } from "../phases/RetreatPhase.js";
import { Command, handleFlags } from "./commands.js";

export class AbstractCombatCommand extends Command {
    constructor() {
        super();
        if (this.constructor === AbstractCombatCommand) {
            throw new Error("AbstractCombatCommand is abstract");
        }
    }

    isDeterministic() {
        return false;
    }

    validateCombat(attackingUnit, attackingHex, defendingUnit, defendingHex) {
        if (!attackingUnit) {
            throw new Error(`No unit at ${attackingHex}`);
        }
        if (!defendingUnit) {
            throw new Error(`No unit at ${defendingHex}`);
        }
        if (attackingUnit.side === defendingUnit.side) {
            throw new Error(`Cannot attack own unit at ${defendingHex}`);
        }
        if (defendingHex === attackingHex) {
            throw new Error(`Cannot attack self at ${defendingHex}`);
        }
    }

    /**
     * @param {Unit} attackingUnit
     * @param {Hex} defendingHex
     * @param {Unit} defendingUnit
     * @param {Game} game
     * @returns {GameEvent[]}
     */
    attack(attackingUnit, defendingHex, defendingUnit, game) {
        let events = [];
        const diceCount = this.decideDiceCount(attackingUnit, game);
        let diceResults = game.roll(diceCount);

        let numberOfFlags = diceResults.filter(r => r === dice.RESULT_FLAG).length;
        let ignorableFlags = game.isSupported(defendingHex) ? 1 : 0;
        let maxDistanceRequired = numberOfFlags * defendingUnit.retreatHexes;
        let retreatPaths = game.retreatPaths(defendingHex, maxDistanceRequired, defendingUnit.side);
        const flagResult = handleFlags(numberOfFlags, defendingUnit.retreatHexes, ignorableFlags, retreatPaths);
        const totalDamage = flagResult.damage +
            defendingUnit.takeDamage(diceResults, this.doesSwordsResultInflictDamage(attackingUnit, defendingUnit));

        game.takeDamage(defendingUnit, totalDamage);
        events.push(new DamageEvent(attackingUnit, defendingUnit, defendingHex, totalDamage, diceResults));
        if (game.isUnitDead(defendingUnit)) {
            events.push(new UnitKilledEvent(defendingHex, defendingUnit));
            if (attackingUnit.side === game.currentSide) {
                game.unshiftPhase(new AdvanceAfterCombatPhase(defendingHex, game.hexOfUnit(attackingUnit)));
            }
        } else if (flagResult.retreats.length > 0) {
            if (attackingUnit.side === game.currentSide) {
                game.unshiftPhase(new AdvanceAfterCombatPhase(defendingHex, game.hexOfUnit(attackingUnit)));
            }
            game.unshiftPhase(new RetreatPhase(defendingUnit.side, defendingHex, flagResult.retreats));
        }
        return events;
    }

    doesSwordsResultInflictDamage(attackingUnit, defendingUnit) {
        throw new Error("Abstract method");
    }

    decideDiceCount(attackingUnit, game) {
        throw new Error("Abstract method");
    }

}
