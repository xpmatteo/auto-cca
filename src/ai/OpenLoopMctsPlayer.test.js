import { fixedRandom, resetFixedRandom } from "../lib/random.js";
import { Command } from "../model/commands/commands.js";
import makeGame from "../model/game.js";
import { TwoOnTwoMeleeScenario } from "../model/scenarios.js";
import { OpenLoopMctsPlayer } from "./OpenLoopMctsPlayer.js";

describe('OpenLoopMctsPlayer', () => {
    test('decideMove returns a valid move', () => {
        const game = makeGame(new TwoOnTwoMeleeScenario());
        const player = new OpenLoopMctsPlayer({
            expansionFactor: 2.1415,
            playoutIterations: 20,
            iterations: 1,
            logfunction: () => {},
        });

        const move = player.decideMove(game);

        expect(move).toBeInstanceOf(Array);
        move.forEach(command => expect(command).toBeInstanceOf(Command));
    });

    const originalRandom = Math.random;
    beforeEach(() => {
        resetFixedRandom();
        Math.random = fixedRandom;
    });

    afterEach(() => {
        Math.random = originalRandom;
    });

    const player = new OpenLoopMctsPlayer({
        expansionFactor: 2.1415,
        playoutIterations: 20,
        logfunction: () => {},
    });

    test('2 on 2', () => {
        const game = makeGame(new TwoOnTwoMeleeScenario());

        const root = player.search(game, 100);

        const rootAsString = root.toString();
        console.log(rootAsString);
        expect(rootAsString).toMatchSnapshot();
    });

});
