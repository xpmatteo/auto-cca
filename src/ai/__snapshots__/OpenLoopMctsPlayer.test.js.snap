// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenLoopMctsPlayer 2 on 2 1`] = `
"undefined -> 27/100: Side: Roman -> 2
 PlayCard(Order Three Units Left) -> 8/40: Side: Roman -> 1
  End phase -> 9/39: Side: Roman -> 11
   Move [0,6] to [1,5] -> 1/3: Side: Roman -> 2
    Move [-1,6] to [-1,7] -> -1/1: Side: Roman -> 0
    Move [-1,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
   Move [0,6] to [0,7] -> 2/4: Side: Roman -> 4
    Move [-1,6] to [-1,7] -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [0,5] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [-1,6] to [-2,6] -> -1/3: Side: Roman -> 3
    Move [0,6] to [-1,7] -> 0/1: Side: Roman -> 0
    Move [0,6] to [-1,6] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    End phase -> -1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Carthaginian -> 0
   End phase -> -1/2: Side: Roman -> 1
    End phase -> 1/2: Side: Carthaginian -> 1
     PlayCard(Order Heavy Troops) -> 0/1: Side: Carthaginian -> 0
   Move [-1,6] to [0,5] -> 3/5: Side: Roman -> 5
    Move [0,6] to [0,7] -> 1/1: Side: Roman -> 0
    End phase -> 1/1: Side: Roman -> 1
     Close Combat from [0,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
    Move [0,6] to [1,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [-1,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [1,5] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [0,6] to [0,5] -> -1/2: Side: Roman -> 2
    Move [-1,6] to [0,6] -> -1/1: Side: Roman -> 0
    Move [-1,6] to [-1,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
   Move [-1,6] to [-1,7] -> 3/5: Side: Roman -> 5
    Move [0,6] to [0,5] -> 1/1: Side: Roman -> 0
    Move [0,6] to [0,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [1,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    End phase -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Carthaginian -> 0
   Move [0,6] to [1,6] -> 3/5: Side: Roman -> 5
    Move [-1,6] to [0,6] -> 1/1: Side: Roman -> 0
    Move [-1,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-1,7] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [-1,6] to [-1,5] -> 1/3: Side: Roman -> 3
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 0
    End phase -> 1/1: Side: Roman -> 1
     Close Combat from [-1,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
    Move [0,6] to [-1,7] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [-1,6] to [-2,7] -> 0/3: Side: Roman -> 3
    Move [0,6] to [-1,7] -> 0/1: Side: Roman -> 0
    End phase -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Carthaginian -> 0
    Move [0,6] to [0,5] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
   Move [0,6] to [-1,7] -> 0/3: Side: Roman -> 3
    End phase -> 0/1: Side: Roman -> 0
    Move [-1,6] to [-2,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
 PlayCard(Order Heavy Troops) -> 19/60: Side: Roman -> 1
  End phase -> 19/60: Side: Roman -> 11
   Move [-1,6] to [-2,7] -> 0/5: Side: Roman -> 4
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 0
    Move [0,6] to [0,7] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
    Move [0,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [1,6] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [0,6] to [1,5] -> -1/3: Side: Roman -> 3
    Move [-1,6] to [0,5] -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,7] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
    Move [-1,6] to [-2,6] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   Move [-1,6] to [0,5] -> 10/11: Side: Roman -> 6
    Move [0,6] to [0,7] -> 2/2: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [1,5] -> 2/2: Side: Roman -> 1
     End phase -> 2/2: Side: Roman -> 1
      Close Combat from [0,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
    End phase -> 1/2: Side: Roman -> 2
     Close Combat from [0,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
     Close Combat from [0,5] to [1,4] -> 0/1: Side: Roman -> 1
      Retreat [0,5] to [-1,6] -> 0/1: Side: Roman -> 0
    Move [0,6] to [1,6] -> 2/2: Side: Roman -> 1
     End phase -> 2/2: Side: Roman -> 1
      Close Combat from [0,5] to [1,4] -> -1/1: Side: Carthaginian -> 0
    Move [0,6] to [-1,6] -> 2/2: Side: Roman -> 1
     End phase -> 2/2: Side: Roman -> 1
      Close Combat from [0,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
    Move [0,6] to [-1,7] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
   Move [0,6] to [0,5] -> 0/4: Side: Roman -> 4
    Move [-1,6] to [0,6] -> -1/1: Side: Roman -> 0
    Move [-1,6] to [-1,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
   Move [-1,6] to [-2,6] -> 0/4: Side: Roman -> 4
    Move [0,6] to [-1,7] -> 0/1: Side: Roman -> 0
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [-1,6] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    Move [0,6] to [0,7] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
   End phase -> -2/3: Side: Roman -> 1
    End phase -> 2/3: Side: Carthaginian -> 1
     PlayCard(Order Heavy Troops) -> 1/2: Side: Carthaginian -> 1
      End phase -> 1/1: Side: Carthaginian -> 0
   Move [-1,6] to [-1,5] -> 4/8: Side: Roman -> 7
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 0
    Move [0,6] to [1,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [0,7] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    End phase -> 1/1: Side: Roman -> 1
     Close Combat from [-1,5] to [0,4] -> -1/1: Side: Carthaginian -> 0
    Move [0,6] to [-1,6] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
    Move [0,6] to [-1,7] -> 0/2: Side: Roman -> 1
     End phase -> 0/2: Side: Roman -> 1
      Close Combat from [-1,5] to [0,4] -> -1/1: Side: Roman -> 0
   Move [-1,6] to [-1,7] -> 1/5: Side: Roman -> 5
    Move [0,6] to [0,7] -> 0/1: Side: Roman -> 0
    Move [0,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [0,6] to [-1,6] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
    Move [0,6] to [1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    End phase -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Carthaginian -> 0
   Move [0,6] to [-1,7] -> 2/5: Side: Roman -> 5
    Move [-1,6] to [-2,7] -> 0/1: Side: Roman -> 0
    Move [-1,6] to [0,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [0,5] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
    End phase -> 1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Carthaginian -> 0
    Move [-1,6] to [-2,6] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
   Move [0,6] to [1,6] -> 3/6: Side: Roman -> 6
    End phase -> -1/1: Side: Roman -> 0
    Move [-1,6] to [0,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [0,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-1,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
   Move [0,6] to [0,7] -> 2/5: Side: Roman -> 5
    Move [-1,6] to [-1,7] -> 1/1: Side: Roman -> 0
    Move [-1,6] to [0,6] -> -1/1: Side: Roman -> 1
     End phase -> -1/1: Side: Roman -> 0
    Move [-1,6] to [-1,5] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,6] -> 1/1: Side: Roman -> 1
     End phase -> 1/1: Side: Roman -> 0
    Move [-1,6] to [-2,7] -> 0/1: Side: Roman -> 1
     End phase -> 0/1: Side: Roman -> 0
"
`;
